swarm_dial.go

```diff
@@ -480,17 +480,37 @@ loop:
 			}
 
 		case <-triggerDial:
-			for _, addr := range nextDial {
+			// we dial batches of addresses together, logically belonging to the same batch
+			// after a batch of addresses has been dialed, we add a delay before initiating the next batch
+			dialed := false
+			last := 0
+			next := 0
+			for i, addr := range nextDial {
+				if dialed && !s.sameAddrBatch(nextDial[last], addr) {
```

Comment:
assuming we actually had delays, wouldn't any new addresses trigger the next group? I think we need to add delays so we can at least reason about this.

Reply:
Assuming we fixed that, I think we'd have a new problem: _better_ dials (e.g., hole punching) might get stuck behind an ongoing "worse" dial group.

Reply:
Not necessarily.

When new addresses come in, we append to existing pending dials and _rerank_ them, so they would get pushed at the top.
At this point we could just trigger a new dial to get them processed.